File: TappyColour/Engine/ConversionSettings.swift
import Foundation

struct ConversionSettings: Equatable, Codable {
    var sliderValue: Float
    var gridSize: Int
    var paletteSize: Int

    static func make(sliderValue: Float) -> ConversionSettings {
        let clamped = max(0.0, min(1.0, sliderValue))
        let gridSize = Int((16 + clamped * Float(80 - 16)).rounded())
        let paletteSize = Int((6 + clamped * Float(24 - 6)).rounded())
        return ConversionSettings(
            sliderValue: clamped,
            gridSize: max(16, min(80, gridSize)),
            paletteSize: max(6, min(24, paletteSize))
        )
    }
}

File: web/src/engine/ConversionSettings.ts
export interface ConversionSettings {
  sliderValue: number
  gridSize: number
  paletteSize: number
  autoFillEnabled: boolean
  imageFit?: 'cover' | 'contain'
}

function clamp(v: number, lo: number, hi: number): number {
  return Math.max(lo, Math.min(hi, v))
}

export function makeConversionSettings(
  sliderValue: number,
  opts: { autoFillEnabled?: boolean; imageFit?: 'cover' | 'contain' } = {}
): ConversionSettings {
  const s = clamp(sliderValue, 0, 1)
  return {
    sliderValue: s,
    gridSize: clamp(Math.round(16 + s * 64), 16, 80),
    paletteSize: clamp(Math.round(6 + s * 18), 6, 24),
    autoFillEnabled: opts.autoFillEnabled ?? true,
    imageFit: opts.imageFit ?? 'cover',
  }
}
